node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/container/ContainerStatsCollector.java

```diff
@@ -1,54 +1,87 @@
     private final CGroup cgroup;
     private final FileSystem fileSystem;
     private final int onlineCpus;
 
-    ContainerStatsCollector(CGroup cgroup, FileSystem fileSystem) {
-        this(cgroup, fileSystem, Runtime.getRuntime().availableProcessors());
+    ContainerStatsCollector(ContainerEngine containerEngine, CGroup cgroup, FileSystem fileSystem) {
+        this(containerEngine, cgroup, fileSystem, Runtime.getRuntime().availableProcessors());
     }
 
-    ContainerStatsCollector(CGroup cgroup, FileSystem fileSystem, int onlineCpus) {
+    ContainerStatsCollector(ContainerEngine containerEngine, CGroup cgroup, FileSystem fileSystem, int onlineCpus) {
+        this.containerEngine = Objects.requireNonNull(containerEngine);
         this.cgroup = Objects.requireNonNull(cgroup);
         this.fileSystem = Objects.requireNonNull(fileSystem);
         this.onlineCpus = onlineCpus;
     }
 
     /** Collect statistics for given container ID and PID */
-    public Optional<ContainerStats> collect(ContainerId containerId, int pid, String iface) {
+    public Optional<ContainerStats> collect(NodeAgentContext context, ContainerId containerId, int pid, String iface) {
         try {
             ContainerStats.CpuStats cpuStats = collectCpuStats(containerId);
             ContainerStats.MemoryStats memoryStats = collectMemoryStats(containerId);
             Map<String, ContainerStats.NetworkStats> networkStats = Map.of(iface, collectNetworkStats(iface, pid));
-            return Optional.of(new ContainerStats(networkStats, memoryStats, cpuStats));
+            List<ContainerStats.GpuStats> gpuStats = collectGpuStats(context);
+            return Optional.of(new ContainerStats(networkStats, memoryStats, cpuStats, gpuStats));
         } catch (NoSuchFileException ignored) {
             return Optional.empty(); // Container disappeared while we collected stats
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }
     }
 
+    private List<ContainerStats.GpuStats> collectGpuStats(NodeAgentContext context) {
+        boolean hasGpu = Files.exists(fileSystem.getPath("/dev/nvidia0"));
+        if (!hasGpu) {
+            return List.of();
+        }
+        Stream<String> lines = containerEngine.execute(context, UnixUser.ROOT, Duration.ofSeconds(5),
+                                                       "nvidia-smi",
+                                                       "--query-gpu=index,utilization.gpu,memory.total,memory.free",
+                                                       "--format=csv,noheader,nounits")
+                                              .getOutputLinesStream();
+        return lines.map(ContainerStatsCollector::parseGpuStats).toList();
+    }
+
+    private static ContainerStats.GpuStats parseGpuStats(String s) {
+        String[] fields = fields(s, ",\\s*");
+        if (fields.length < 4) throw new IllegalArgumentException("Could not parse GPU stats from '" + s + "'");
```

Comment:
Throw `ConvergenceException.ofError()` instead

Reply:
This is in the "should never happen" class of exceptions, so leaving it as-is.